Redesign

Somethingweirdhere opened the pull request:
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.

First part of my redesign.

I added a navigation drawer with all things we need there.
This isn't done yet, I still need to remove the options menu, add a better navigation and add the tabs.

Also, we might want to get different icons for the different kiosks.

Somethingweirdhere commented:
Seems pretty much done to me. I will keep testing for bugs.

TobiGr commented:
I just installed it and it really works like a charm! Here's some feedback after a short test:

With the light theme the icon colors do not fit to the original ones: [screenshot](https://user-images.githubusercontent.com/17365767/41227614-c64d4174-6d75-11e8-8055-4a5a19439c99.png)

Adding support for different themes to your new `Content of main page` setting would be nice too.

@Somethingweirdhere @theScrabi @karyogamy What do you think of removing a main page type from the `Selection` list when it's added to the `Your tabs` list? 
I think this is quite useful for two reasons: I guess nobody wants to add multiple blank or history pages to their main page. Secondly this would allow the `Your tabs` section to increase its height. That's good for small phones (like mine) because I can only see two to three entries without scrolling.
Maybe we can differentiate between _single_ (blank, subscriptions, feed, bookmarks, history) and _multi_ (kiosk, channel) use main page types.

Somethingweirdhere commented:
@TobiGr True, forgot about that! I will add theme support soon.

Also yeah sounds like a good idea, I can't think of an use for multiple of these tabs. Maybe someone more creative can?  Also kiosks are only used once, if you switch service they adjust on their own. So the only multi page would be the channel.

Somethingweirdhere commented:
@TobiGr Alright, theme support done.

About removing the tabs, let's wait for the opinions of others and see what they think.

Thanks for the feedback :)

theScrabi commented:
I can take a look at this next weekend. Could someone please paste some screnshots :)

Somethingweirdhere commented:
@theScrabi Sure, I hope you don't mind if they are in german.

![screenshot_20180612-211451](https://user-images.githubusercontent.com/36226087/41311888-f26a7446-6e85-11e8-931e-2d0cf08e139b.jpg)
![screenshot_20180612-211506](https://user-images.githubusercontent.com/36226087/41311889-f28bd0d2-6e85-11e8-901d-f85b92aa62d4.jpg)
![Uploading Screenshot_20180612-211517.jpg…]()
![Uploading Screenshot_20180612-211523.jpg…]()
![Uploading Screenshot_20180612-211526.jpg…]()

theScrabi commented:
![serveimage](https://user-images.githubusercontent.com/1891273/41313137-67ae24e8-6e89-11e8-8843-e71d5404e981.jpeg)

theScrabi commented:
By the way if you are already working on it, could you fix the Navigation drawer UI, so it slips below the title bar?

It's a little detail, but you know what they said in Westworld. People are not coming back for the big thing, they are coming back for the details ;D

theScrabi commented:
@TobiGr I would not remove the tabs, but I see your problem.

Currently the layout is some sort of "hack", we use two action bars that look like one. Maybe the bottom one could be hidden if only one tab is visible.

theScrabi commented:
@Somethingweirdhere 
However the one thing I don't like is that the services are top level with the selection of Kiosk, feed, etc. If we only have two services its OK, but this number will vary, and definitely increase.
I don't know how much services we will support, but I would make the UI be able to handle as much services as possible.

theScrabi commented:
Also this thing here :D [image](https://user-images.githubusercontent.com/36226087/41311930-0cf86f20-6e86-11e8-821d-771ff82df849.jpg). I would merge it already, but it could take a litte more of a "materialish" touch.

karyogamy commented:
> However the one thing I don't like is that the services are top level with the selection of Kiosk, feed, etc. If we only have two services its OK, but this number will vary, and definitely increase.

What I was about to say as well. The selection dialog should be turned into a separate settings fragment and use recyclerviews for drag and dropping. We can also add a user-defined selection of services at the same place as well.

Somethingweirdhere commented:
@theScrabi I can do the drawer thing, if you can show me what you mean. I have taken a look at Gmail and the behaviour seems the same. Can you maybe show me a Gif or give me an app name?

And yeah you are right, at some point it would be a lot of scrolling if there are more services. Any idea where I should put them? Between History and Settings? Or should we deal with the issue as soon as we have more services? 

I don't think adding an expandable service section is useful right now, so maybe let's wait with that.

And yeah I 100% agree that dialog looks horrible, but that's my best in terms of design. I really hope someone who enjoys designing can pick some better colors and icons.

theScrabi commented:
@Somethingweirdhere maybe you can do something like we did with our owncloud app:
https://github.com/owncloud/android/issues/2204#issuecomment-386546316

You have this triangle menu there in the header. If you tab the header you can change the mode of the navigation bar, and you would see different elements.

So in the owncloud app normally you would see the selections for files, actions, and the settings, but if you tab in the header you can select a user. Now for us instead of selecting a user one could select a service.

theScrabi commented:
This is what I mean with the drawer below the status bar:

#### Do
![ky6vj](https://user-images.githubusercontent.com/1891273/41317973-132ea470-6e97-11e8-82af-98e82c7776e5.png)

#### Don't do
![serveimage](https://user-images.githubusercontent.com/1891273/41317982-1c88c41a-6e97-11e8-8adb-66a187c1585d.jpeg)

theScrabi commented:
@TobiGr I thought about the problem with the tabs. I think the tab bar might also be be collapsible like this example here at twidere.

![out](https://user-images.githubusercontent.com/1891273/41346435-bfe74202-6f06-11e8-9b99-3220671451b9.gif)

However the search bar is definetly not collapsible.

TobiGr commented:
@theScrabi I am not sure if I understood your answer correctly, but I meant the settings dialog you linked here: https://github.com/TeamNewPipe/NewPipe/pull/1461#issuecomment-396711955
In my comment I was referring to the "Your tabs" RecyclerView. 

> What I was about to say as well. The selection dialog should be turned into a separate settings fragment and use recyclerviews for drag and dropping. We can also add a user-defined selection of services at the same place as well.

That sounds good.

Somethingweirdhere commented:
Alright, first thing is done. This took way to long to figure out (about 5 hours) for this tiny effect. I really hope this is a gamechanger for someone 😂

Somethingweirdhere commented:
The OwnCloud-like menu now works too. I can't say that I like this implementation (the switching feels very abrupt), but I don't have any better idea.

Now, the last thing I have to fix is this the selection dialog, right? Does what karyogamy said sound good to you @theScrabi ? A completely new settings fragment would mean there is enough space, so we don't need to remove already used tabs, right?

![screenshot_20180614-033422](https://user-images.githubusercontent.com/36226087/41386854-071820a6-6f84-11e8-9882-49cc83ffea24.jpg)
![screenshot_20180614-033427](https://user-images.githubusercontent.com/36226087/41386855-0736153e-6f84-11e8-8a08-8889f46075aa.jpg)

theScrabi commented:
> This took way to long to figure out (about 5 hours) for this tiny effect. I really hope this is a gamechanger for someone 😂

I know it's the tiny changes that take the most time, but I really appreciate it :D. It makes the app look more elaborate, and that is good for UX. I'd say you didn't waste the 5 hours.

Somethingweirdhere commented:
I did this

`The selection dialog should be turned into a separate settings fragment and use recyclerviews for drag and dropping`

Still not perfect, but imo it looks mergeable now. What do you think?

![screenshot_20180615-052537](https://user-images.githubusercontent.com/36226087/41448745-c7f29fb2-705c-11e8-9419-49acaa9b1ecd.jpg)

![image](https://user-images.githubusercontent.com/36226087/41448792-284e412c-705d-11e8-8ed4-5ef8d962b90e.png)

karyogamy commented on app/src/main/java/org/schabi/newpipe/MainActivity.java:
```diff
@@ -83,28 +98,64 @@ protected void onCreate(Bundle savedInstanceState) {
 
-        drawerItems.getMenu().getItem(ServiceHelper.getSelectedServiceId(this)).setChecked(true);
+        drawerItems.getMenu()
+                .add(R.id.menu_tabs_group, -1, 0, R.string.tab_subscriptions)
```
These index numbers should be extracted into static variables, since they are used multiple times.

karyogamy commented on app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java:
```diff
@@ -59,56 +82,145 @@
+                    }
+                } else {
+                    switch (tabNumber) {
+                        case "0":
```
Similar problem as above.

karyogamy commented on app/src/main/java/org/schabi/newpipe/local/subscription/SubscriptionFragment.java:
```diff
@@ -239,8 +240,15 @@ private void setupExportToItems(final ViewGroup listHolder) {
-        NavigationHelper.openSubscriptionsImportFragment(getParentFragment().getFragmentManager(), serviceId);
+        FragmentManager fragmentManager;
+
+        if (getParentFragment() == null)
```
This code block is repeated three times in the same file, please extract them into the same method.

karyogamy commented:
Thanks. 

UI-wise, I think the font is a bit too large and the "Your tab ... +" line is a bit unclear. Maybe this should be at the bottom of the list and changed to "+ New Tab". Also, a title in the dialog would help clarifying things as well.

I've also quickly glanced over the code and saw a few things that might need addressing, but they are mostly for polishing right before merge. Though with anything related to the navigation drawer logic, please wait for @theScrabi to do a more in-depth code review, since he authored most of the original code there.

Somethingweirdhere commented:
@karyogamy Thanks 👍 

A title is a good idea, but I can't decide how to place the new tab line. I could make it look exactly like the tabs, and just make sure it won't move. Or make it look special/bigger? Or put it a bit under the rest of the tabs? What do you think? And should it be activated by pressing anywhere on the line, or by pressing on the + next to it?

Font-size smaller, but viewholdersize should stay the same, right?

Also, I hope the navigation by long-pressing to move and sliding to remove is fine.

karyogamy commented:
> Or make it look special/bigger?

It seems to have a smaller font on the screenshot you posted, so it might be a good idea to have the same font size as the rest of the text. Since its the only line with an icon at the front, it's probably special enough =D

> Or put it a bit under the rest of the tabs?

Please do.

> And should it be activated by pressing anywhere on the line, or by pressing on the + next to it?

The whole line please, this way both hands can reach the button.

> Font-size smaller, but viewholdersize should stay the same, right?

Yep. I don't think the font size in the fragment should exceed the font size of the text on the app bar at the top.

> Also, I hope the navigation by long-pressing to move and sliding to remove is fine.

I'm fine with that, though ideally we would have a handle for dragging so the user knows it can be done.

Somethingweirdhere commented:
1. Alright.
2. Good idea
3. Can do, but I personally disagree with that. Changing Services now changes the kiosks accordingly. Being able to select Services, and the tab color changing makes sense this way. But if the only thing that Changes when selecting Services is the avaible selection of kiosks in the navigation drawer and the search, it might be better to just add all avaible Services in the drawer in drop-down menus or something like this and adding a Switch for each Service in the search menu. 

It's your choice of coutse, but I dont feel like it makes sense that the tabs dont change with the services.
4. Confirmed, I need to set the view size to matchparent.

Somethingweirdhere commented:
Alright, cleaned up the code, rebased (at least I think I did?), placed to appearance, fixed the bug, made the text smaller, used cardview, made add tab at the bottom and made the dialog look a bit nicer.

![screenshot_20180616-131428](https://user-images.githubusercontent.com/36226087/41498200-a74377d0-7167-11e8-8bbf-ab3f4a447970.jpg)
![screenshot_20180616-131433](https://user-images.githubusercontent.com/36226087/41498201-a75eeff6-7167-11e8-80af-f25f8f7a552b.jpg)
![screenshot_20180616-131437](https://user-images.githubusercontent.com/36226087/41498202-a779a13e-7167-11e8-9ffc-49a02d2d9292.jpg)

Somethingweirdhere commented:
@TobiGr Thank you :) Yeah you have good suggestions, I'm bad with UI design but it definitely looks better now:

![image](https://user-images.githubusercontent.com/36226087/41512211-9492fae2-7285-11e8-96e0-d14147ac1332.png)
![image](https://user-images.githubusercontent.com/36226087/41512229-9b600090-7285-11e8-9dee-b58e676ff499.png)

Everything is a bit bigger and easier to touch 👍 

I agree the + New Tab isn't the best way, but I also don't think a floating action button is any better, and I don't think it fit's the app. But I also have no better ideas :/

And yeah, it sounds like a good idea to put some info out there. But not all users read the blog, and it might be useful to put something in the app that informs them? They will notice the drawer anyways, but I doubt many will look into the appearance settings unless they know this setting exists.

theScrabi commented:
@Somethingweirdhere nice :D we are getting close.

Here are some finetunigs:
- For the cards view I would suggest to also increase the margin a little bit, so they actually look like cards and can be handled as such.
- Maybe if the cards are a bit higher it might also be usefull.
- The + card on the bottom is ok, but a FAB button might be the better solutions.
- I think it was good to add a handle to the right side of the cards, so it is clear for the user that they can move them. Like this here:
![screenshot_20180618-101045](https://user-images.githubusercontent.com/1891273/41525117-3ae0cb68-72e0-11e8-9dde-5d778d37beac.png)

Somethingweirdhere commented:
![image](https://user-images.githubusercontent.com/36226087/41528852-970d384e-72eb-11e8-8e81-08cc771df636.png)

Alright, that looks quite good to me!

theScrabi commented:
@Somethingweirdhere verry nice :)

Somethingweirdhere commented:
Thanks :) 

 - Alright, using vector graphics now.
 - Set to 4dp
 - Thought that was intentional? 

styles.xml:

    <!-- Settings -->
    <style name="LightSettingsTheme" parent="LightTheme">
        <item name="colorAccent">@color/light_settings_accent_color</item>
    </style>

    <style name="DarkSettingsTheme" parent="DarkTheme">
        <item name="colorAccent">@color/dark_settings_accent_color</item>
    </style>

    <style name="BlackSettingsTheme" parent="BlackTheme">
        <item name="colorAccent">@color/black_settings_accent_color</item>
    </style>

I will change it for the button only then, but maybe we should change it in the styles too?

Anyways, here it is :)

![image](https://user-images.githubusercontent.com/36226087/41550745-101c5c22-732a-11e8-8850-bf71df8f4669.png)

theScrabi commented:
![704](https://user-images.githubusercontent.com/1891273/41556099-fc785756-7339-11e8-9648-0c9e55a54b2d.jpg)

Ok so I'm happy with that part of the UI. Good job :)

TobiGr commented:
> I think it was better if we have one global layout that applies to all services.

@theScrabi I really like the way it is right now, because every service is different. E.g YouTube has only one kiosk (trending) and SoundCloud provides two. Additionally, the "What's New" page is weight more useful with YouTube than with SoundCloud. At least the way I used NewPipe, it has shown me that new media are created on YouTube much more often than on SoundCloud.

Somethingweirdhere commented:
@theScrabi I have told you about it  to adress this: `4. The actuall Kisok should be selectable when you add the kiosk tab (like you do it with the channel tap)`

In short: It feels weird that choosing a service doesn't also change the kiosk tabs. 

Having only one layout that stays would mean that you have to open the drawer two times just to change kiosks if they are from different services. 
Having two layouts means that if you have a kiosk tab, you just need to go to it. 
But if there is only one layout, I don't see why we are putting the services in a seperate container, that you need to open after opening the drawer. And then it doesn't change anything but the drawer itself and the search, but not the actual content that is displayed then.

If you want one layout, it would make more sense to put an option to change the service into the search. And the option to select the kiosk in a dropdown menu below the avaible services, instead of having to change those services. Because with one layout for all services, there is no reason to change services, except for searching and opening a differnt kiosk.

I hope you understand what I mean. I just wanted to share my thoughts. If you disagree with me, I will change it back to one layout and ask a kiosk selection.

theScrabi commented:
> If you want one layout, it would make more sense to put an option to change the service into the search.

I get your point, and there are more people who suggested this before. Maybe it makes sense to put the service selection near search one day. You are right, the tabs are mostly service independent. The content on them should not change if you select a different service, and I aggree that it is not the best design, however Imagine you have 20 services, and for every service you have to edit the layout again. I think this is getting quite exhausting. I think most users aren't interested in more than a few Kiosks maybe (for youtube no one is actually interested in their bullshit trending). So yes one layout for all services should do it.

TobiGr commented:
@theScrabi Oh, I missed your second paragraph :D 

> Also I didn't like that all the different kiosks available for a service are still in the drawer. I have the fear that this will escalate if a service has to many of them.

I agree.

> Last but not least It's still not possible to select the specific kiosk if you want to add a kiosk tab:

I'd say, we can skip the selection dialog if there is only one kiosk available. But in cases like SoundCloud I'd like to be able to select the kiosk too.

theScrabi commented:
> 
> > Also I didn't like that all the different kiosks available for a service are still in the drawer. I have the fear that this will escalate if a service has to many of them.
>
>I agree.

So I thought about this once more, and I think for now the layout for selecting kiosks in the drawer is quite enough, we just had to change it later this year. Maybe to something like this:
![screenshot_20180618-214047](https://user-images.githubusercontent.com/1891273/41559440-529dbc20-7344-11e8-9b21-0cce7ee8bee4.png)

TobiGr commented:
Oh stop. I misread your text.
I really like the kiosk in the drawer. And if they become too many, we can simply put them between the settings and the history and add an extra divider. Nevertheless, I think we can wait with a final decision on that. I'd wonder what @karyogamy thinks about that.

Somethingweirdhere commented:
@theScrabi I think that's what you mean, right? I have to disagree with this though: `Imagine you have 20 services, and for every service you have to edit the layout again.`

No, I made sure that you just need to add a service, and the layout will work. If there is no kiosk, the kiosk tabs will disappear, if there are 3 kiosks, then all 3 will be inserted. There is nothing more to do than adding the service.

Anyways, here is the commit, 2d2d5a042c124338030b3bec43ff2a470ea8c8a3 is the one with my preffered implementation. Feel free to test both, and before merging I would also be interested what @karyogamy thinks about this.

karyogamy commented:
Ideally, I think we should have a setting fragment to control the behaviour of each service. This way:

- On the viewpager, we can have in the settings fragment the option toggle on and off the kiosk tab that is available on the pager for each service. This way, we have no need to specify individual kiosk in the main page content settings (so we won't have youtube's trending page when soundcloud is selected) and the relative position of the kiosk tabs is maintained. 

- On the drawers, we can just add an expand/collapse button when the number of kiosk is over some limit and/or when certain kiosks are toggled off in the settings.

However, this is probably a lot of work implementation-wise. So for now, with only two services and only two kiosks for soundcloud, I suggest keeping the services with their own kiosks rather than setting the kiosks individually. We can have a more elaborate control later down the road when the need comes. 

Nice job on the UI by the way, I think it is a good time we start using cards to follow the new material design 2.0 guidelines.

Somethingweirdhere commented:
@karyogamy Thanks for you feedback! 😄 

> On the drawers, we can just add an expand/collapse button when the number of kiosk is over some limit and/or when certain kiosks are toggled off in the settings.

I agree we need to do this someday, when we have more services. 

I'm not sure if I understood correctly, but which approach do you prefer for now? Chosen kiosks, or changing ones?

karyogamy commented:
> I'm not sure if I understood correctly, but which approach do you prefer for now? Chosen kiosks, or changing ones?

Changing ones, so the soundcloud kiosks get a fair chance of showing up when the user switches service. Right now they are just buried in a nested menu. Though we might want to do something about the viewpager icons, since both kiosks have the same one and can be confusing.

Somethingweirdhere commented:
Alright, thanks!
Yeah, we definitely need two new icons anyways, as they are also being used in the drawer.

theScrabi commented:
> Changing ones, so the soundcloud kiosks get a fair chance of showing up when the user switches service. 

Yes I agree I would prefer this as well.

> On the drawers, we can just add an expand/collapse button when the number of kiosk is over some limit and/or when certain kiosks are toggled off in the settings.

Sounds good. We should do this once we have more services.


Ok so I think we can agree that we leave the tap implementation that @Somethingweirdhere created, where you __don't__ get to select a specific Kiosk, but in addition the kiosk changes depending on the selected service. (I hope I understood this right)

Here it would also be good if the SoundCloud kiosks get fixed. @wb9688 how are you, still busy? :D

theScrabi commented:
> Alright, sounds good. Can you merge a commit, or do you need me to revert the last 2 commits before?

Please revert the two commits.

> And anything else that should be changed in this PR?

I think the UI should be good by then, but we need to see. I'm currently field testing you changes (I use the app with your PR), also I have to review the code. I can't tell yet if there should be something changed, but I doubt that there will be big changes if the UI is fine then :)

theScrabi commented:
Ok so third test round reveled some more bugs:

1. On android 4.2 the triangle in the drawer glues to the right side:
![screen](https://user-images.githubusercontent.com/1891273/41592805-26f06170-73be-11e8-9812-b3637e2f16e2.png)

Somethingweirdhere commented:
1. Fixed

2. Fixed too, but the dialog looks different. I think that's normal for those versions. (And how all our dialogs look there)

3. I think it's done, but I will have to test with more screen sizes:

![image](https://user-images.githubusercontent.com/36226087/41595531-64b0697a-73c7-11e8-897f-de68f37407fb.png)


4. I tried using the same thing that Watchlists/bookmarks are using, but couldn't manage to implement that. I assume you mean that you should be able to move around the cards when pressing on the handle, and if you long press they should be removed?

theScrabi commented:
> 2. Fixed too, but the dialog looks different. I think that's normal for those versions. (And how all our dialogs look there)

Honestly I don't care how it looks on old devices, as long as it works :P

Somethingweirdhere commented:
Alright, it worked :)

theScrabi commented on app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java:
```diff
@@ -59,56 +90,145 @@
-    /*//////////////////////////////////////////////////////////////////////////
-    // Menu
-    //////////////////////////////////////////////////////////////////////////*/
+    private void setIcons() {
```
Cant the icons and the titles be setup simultaniously? This looks like redundant code. Maybe you should consider a structure where you call ´setupTabs()` where you don't have to write the switch case etc. twice.

theScrabi commented on app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java:
```diff
@@ -59,56 +90,145 @@
+                    if (kiosk.length == 2) {
+                        try {
+                            tabToSet.setIcon(KioskTranslator.getKioskIcons(kiosk[1], getContext()));
+                        } catch (Exception e) {
```
This is considered bad coding. What is the Exception thrown here, and could you prevent it?

theScrabi commented on app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java:
```diff
@@ -157,120 +300,74 @@ public void onTabReselected(TabLayout.Tab tab) {
-                        return new SubscriptionFragment();
+            String tabNumber = tabs.get(position);
+
+            if(tabNumber.startsWith(TAB_NUMBER_KIOSK + "\t")) {
```
What is tabNumber, and why does it need to be split later?
What would be a typical value it holds?

theScrabi commented on app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java:
```diff
@@ -157,120 +300,74 @@ public void onTabReselected(TabLayout.Tab tab) {
-            i++;
+        @Override
+        public void destroyItem(ViewGroup container, int position, Object object) {
+            getFragmentManager().beginTransaction().remove((Fragment)object).commitNowAllowingStateLoss();
```
You should write this multiline:
```
getFragmentManager()
    .beginTransaction()
    .remove((Fragment)object)
    .commitNowAllowingStateLoss();
```
Makes it more readable ;)

theScrabi commented on app/src/main/java/org/schabi/newpipe/fragments/list/channel/ChannelFragment.java:
```diff
@@ -90,6 +90,8 @@
 
     private MenuItem menuRssButton;
 
+    private boolean mIsVisibleToUser = false;
```
See the commend below.

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/bookmark/BookmarkFragment.java:
```diff
@@ -66,11 +66,10 @@ public void onCreate(Bundle savedInstanceState) {
-            activity.setTitle(R.string.tab_subscriptions);
-        }
 
+        if(!useAsFrontPage) {
```
Same as with mIsVisibleToUser. This should maybe be put into `BaseFrament` then onCreateView() should call its super.

theScrabi replied:
then onCreateView() should call its super.

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -71,6 +71,10 @@ public void onCreate(Bundle savedInstanceState) {
     @Override
     public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
+
+        if(!useAsFrontPage) {
```
Same as with mIsVisibleToUser. This should maybe be put into `BaseFrament` then onCreateView() should call its super.

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -105,6 +109,14 @@ public void onDestroyView() {
+    }
+
     /*@Override
     protected RecyclerView.LayoutManager getListLayoutManager() {
```
Remove this on the way :)

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/feed/FeedFragment.java:
```diff
@@ -105,6 +109,14 @@ public void onDestroyView() {
     }
 
+    @Override
+    public void setUserVisibleHint(boolean isVisibleToUser) {
```
Redundant code: Same as with mIsVisibleToUser. This should maybe be put into `BaseFrament`.

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/history/StatisticsPlaylistFragment.java:
```diff
@@ -96,14 +97,24 @@ public View onCreateView(@NonNull LayoutInflater inflater,
     }
 
+    @Override
+    public void setUserVisibleHint(boolean isVisibleToUser) {
```
Redundant code: Same as with mIsVisibleToUser. This should maybe be put into `BaseFrament`.

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/history/StatisticsPlaylistFragment.java:
```diff
@@ -96,14 +97,24 @@ public View onCreateView(@NonNull LayoutInflater inflater,
     protected void initViews(View rootView, Bundle savedInstanceState) {
         super.initViews(rootView, savedInstanceState);
-        setTitle(getString(R.string.title_last_played));
+        if(!useAsFrontPage) {
```
Redundant code: Same as with mIsVisibleToUser. This should maybe be put into `BaseFrament`.

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/history/StatisticsPlaylistFragment.java:
```diff
@@ -129,8 +140,8 @@ protected void initListeners() {
                     final StreamStatisticsEntry item = (StreamStatisticsEntry) selectedItem;
-                    NavigationHelper.openVideoDetailFragment(getFragmentManager(),
-                            item.serviceId, item.url, item.title);
+                    NavigationHelper.openVideoDetailFragment(useAsFrontPage ? getParentFragment().getFragmentManager() : getFragmentManager(), item.serviceId, item.url, item.title);
```
Write it like this:
```
NavigationHelper.openVideoDetailFragment(useAsFrontPage
        ? getParentFragment().getFragmentManager()
        : getFragmentManager(),
    item.serviceId,
    item.url,
    item.title);
```

Makes it more readable.

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/subscription/SubscriptionFragment.java:
```diff
@@ -318,18 +319,30 @@ protected void initListeners() {
         //noinspection ConstantConditions
         whatsNewItemListHeader.setOnClickListener(v ->
-                NavigationHelper.openWhatsNewFragment(getParentFragment().getFragmentManager()));
+        {
```
Put the bracket into line 329.

theScrabi commented on app/src/main/java/org/schabi/newpipe/local/subscription/SubscriptionFragment.java:
```diff
@@ -318,18 +319,30 @@ protected void initListeners() {
 
+    private FragmentManager getFM() {
+        if (getParentFragment() == null)
+        {
```
Puth the braket into line 338

theScrabi replied:
also redundant code.

theScrabi commented on app/src/main/java/org/schabi/newpipe/settings/ContentSettingsMain.java:
```diff
@@ -0,0 +1,288 @@
+        }
+    }
+
+    public class UsedAdapter extends RecyclerView.Adapter<ContentSettingsMain.UsedAdapter.TabViewHolder>{
```
Choose a name that clarifies what this adapter does, UsedAdapter is to vague.

theScrabi commented on app/src/main/java/org/schabi/newpipe/settings/ContentSettingsMain.java:
```diff
@@ -0,0 +1,288 @@
+    }
+
+    public class UsedAdapter extends RecyclerView.Adapter<ContentSettingsMain.UsedAdapter.TabViewHolder>{
+        // ... code from gist
```
link to gist pls

theScrabi commented on app/src/main/java/org/schabi/newpipe/settings/ContentSettingsMain.java:
```diff
@@ -0,0 +1,288 @@
+import java.util.Collections;
+import java.util.List;
+
+public class ContentSettingsMain extends Fragment {
```
I suppose this is the fragment with which I can setup the tabs right? It should have a name that clearly depicts that it is responsible for that.

theScrabi commented on app/src/main/java/org/schabi/newpipe/settings/ContentSettingsMainDialog.java:
```diff
@@ -0,0 +1,123 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class ContentSettingsMainDialog extends DialogFragment {
```
Remove this whole class and use `AlertDialog` instead, like [here](https://github.com/teamnewpipe/NewPipe/blob/38946e4b0f28501fd357b3d4f894646aa7d49146/app/src/main/java/org/schabi/newpipe/info_list/InfoItemDialog.java#L45).

theScrabi commented on app/src/main/res/layout/dialog_contentsettingtab.xml:
```diff
@@ -0,0 +1,40 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Remove this and use AlertDialog.

theScrabi commented on app/src/main/res/layout/dialog_contentsettingsadd.xml:
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="utf-8"?>
```
Remove this and use AlertDialog

theScrabi commented on app/src/main/res/layout/dialog_contentsettings.xml:
```diff
@@ -0,0 +1,32 @@
+<?xml version="1.0" encoding="utf-8"?>
+<RelativeLayout xmlns:android="http://schemas.android.com/apk/res/android"
```
Give this resource a better name. It's not a dialog, its a fragment, and it should somehow have the "tab" in its name. "contentSettings" is to generic.

Somethingweirdhere commented:
Alright, this should be better! You are right, I definitely need some experience, I started learning java only a year ago, and except for an udacity course mainly from code snippets found on stackoverflow. So yeah I never really coded with any guidelines other than "it compiles" before.

Anyways, to the review:

I fixed everything that's not related to the BaseFragment. I only put getFM() there, as it's the only function introduced in this PR. For the rest, I guess I will wait for @karyogamy's opinion, and/or do it in another PR afterwards.

The exceptions you mentioned also shouldn't be thrown anymore, but I will need to do testing again to see if I broke anything important in the last commit. For now, I have found no issues.

I also replaced the Dialog with an AlertDialog and renamed some stuff.

About making one function from setIcon and setFirstTitle:
If you take a closer look, you will see that the title is only set for the first Tab `String tabNumber = tabs.get(0);` , while the icons are set for all tabs. I can put them into one function, but then in each case I will need to do `if (i == 0) setTitle(xxx)` [ ] What do you think?

You are right, tabNumber isn't the best name for a string. I renamed it to tabInformation as it contains information about the tab, such as which tab it is, and incase of channels which Url and Name it is.

And last, I have a question about this writing style:

`getFragmentManager()
    .beginTransaction()
    .remove((Fragment)object)
    .commitNowAllowingStateLoss();`

When do I do it, when don't I? For example, 

1. 
`View view = findViewById(R.id.tV);
view.setVisibility(gone);`
2.
`findViewById(R.id.tV).setVisibility(gone)`
3.
`findViewById(R.id.tV)
    .setVisibility(gone)`

If I understand correctly, it's 2.? So do it in one line if only the final command does something while the ones before only provide the class, and split it up if every command does something?

theScrabi commented:
> About making one function from setIcon and setFirstTitle:
If you take a closer look, you will see that the title is only set for the first Tab `String tabNumber = tabs.get(0);` , while the icons are set for all tabs. I can put them into one function, but then in each case I will need to do `if (i == 0) setTitle(xxx)` What do you think? Imo it would make it look more confusing.

I see. That wouldn't make much sense. I'm trying to get a closer look on this.

Somethingweirdhere commented:
Alright, thanks for cleaning up my mess :) 

Tell me when you have decided what to do, meanwhile I will do some cleaning work on my other PR.

theScrabi commented:
@Somethingweirdhere I have another change request for the UI. Removing the main menu might not be so good at the end because it's quite uncomfortable if you have to move to the main screen every time you want to enter for example the settings.

Please add these menues to the main menu again:
- Downloads
- History
- Settings

![screenshot_20180625-132650](https://user-images.githubusercontent.com/1891273/41853144-5a722ea6-788d-11e8-8cc3-f9009baa8e63.png)

Somethingweirdhere commented:
Is this what you mean?

theScrabi commented:
@Somethingweirdhere if you want to continue, the issue with the changing trending tabs hat to be fixed: 
![screenshot_20180628-115224](https://user-images.githubusercontent.com/1891273/42028837-c8ac3064-7acd-11e8-98b7-5769bbe95102.png)

Im still a fan of selecting a prefaced kiosk for the tab view, since the tab selection for the current service can be done in the sidebar, and at least this (I think) would be the solution with the lowest effort right now.

![screenshot_20180628-122420](https://user-images.githubusercontent.com/1891273/42029402-bd3408e0-7acf-11e8-9eee-2a454b04994a.png)


I recommend to think about another trending solution in another issue/pull. Lets first try to get this one here done.

Somethingweirdhere commented:
Alright, I have reverted the reverts, I hope it didn't break anything.

theScrabi commented:
@Somethingweirdhere sorry for that late response, whenever I found time I would work on the back end.

About the taps eddit fragment, I think there was a misunderstanding. I dind't mean to delete the card directly on a long click, but rather unlock the left/right slide on a long click. So you would be able to up/down slide them when only touchen, and left right when longclicking.

### How it's atm.
![gurken](https://user-images.githubusercontent.com/1891273/42450802-a69bf87c-8385-11e8-8beb-20813a35e5e9.gif)

### How I thought it should be
![gurken2](https://user-images.githubusercontent.com/1891273/42450915-f82738fa-8385-11e8-9150-90f2f5035731.gif)

Now I remember you said something like this behavior as I described it was not possible, than just rewind it to how it was before, with long click for up/down, and short for right/left.

Somethingweirdhere commented:
Don't worry about it.

Anyways, I implemented it, but it doesn't feel very intuitive to me.
Please try it out and tell me what you think.

theScrabi commented on app/src/main/java/org/schabi/newpipe/fragments/MainFragment.java:
```diff
@@ -38,65 +44,172 @@
+    private static final String TAB_NUMBER_HISTORY = "5";
+    private static final String TAB_NUMBER_CHANNEL = "6";
+
+    SharedPreferences.OnSharedPreferenceChangeListener listener = (prefs, key) -> {
```
Please put this lambda into line 84, and remove `listener`.

theScrabi commented:
> I implemented it, but it doesn't feel very intuitive to me.

Thank you :) Lets see, if it turns out not to be that good we can change it later :)

Somethingweirdhere commented:
Alright, done :)

theScrabi commented:
I currently need to work, however I have some spare time in two weeks. Then I will work on what is still left.

Somethingweirdhere commented:
@theSchabi Alright, I should have some spare time these days too, would be great to finish this :)

theScrabi commented:
@Somethingweirdhere nice :). Please try to rebase with dev. I'll try to take over from there.

Somethingweirdhere commented:
Alright, I resolved the conflicts. 

I don't see a button to rebase here, I hope merging was fine too?

theScrabi commented:
> I don't see a button to rebase here, I hope merging was fine too?

Merge is not a good idea, because you will see that `dev` branch now has two lines going into you're branch. A rebase however changes the commit from which you originally forked your branch to the current commit of `dev` this makes the history better readable, otherwise you will endup with a web of merges.

Rebase can only be done locally:
```
git checkout dev
git pull
git checkout <your_branch>
git rebase dev
```
done.

Somethingweirdhere commented:
Alright, thanks for explaining, I will try it out next time :)

Is there anything else I should do?

TobiGr commented on app/src/main/res/values/strings.xml:
```diff
@@ -357,6 +359,9 @@
     <string name="main_page_content">Content of main page</string>
+    <string name="main_page_content_summary">What tabs are shown on the main page</string>
+    <string name="selection">Selection</string>
+    <string name="chosenTabs">Your tabs</string>
```
Is there a reason why you used a capital letter instead an underscore?

Somethingweirdhere replied:
Nope, fixing it right now.

TobiGr commented:
I've been using this branch for a couple of weeks now, and I could not detect any bugs. So in my eyes this is good to go, but I do not know if @theScrabi intends to do some more code review which I do not have time for at the moment.

Somethingweirdhere commented:
@TobiGr Thanks for testing :) I'm happy it worked well for you. 

Thanks for pointing out the string name. I have checked and it's unused anyways, so I removed it. We still have 2577 unused resources, so I would suggest that I make a new PR that cleans up a bit later on? Or is there a reason to keep unused resources?

TobiGr commented:
I already went through the resources two weeks ago and tried to use Android Studio's build in "remove unused resources" tool. Unfortunately this doesn't worked well since it does not find every occurrence of the resources. Therefore I wrote a small tool which does a more reliable job. I think I am going to open that clean up PR next week. See #1603.

theScrabi commented:
I rebased and merged it :).
Somehow github does not show this branch is merged, so I have to close it :/

@Somethingweirdhere verry nice work :+1:

Somethingweirdhere commented:
The result is the same, so whatever^^

By the way, you also closed the other PR I made to fix long links that weren't clickable correctly. What's up with that? It still doesn't seem to work.

mauriciocolli commented:
This pull request had several things to fix before merge, that (somehow) slipped through the review. We have to consider fixing them at least before releasing, though the issues are mostly code style and bad practices.

> Somehow github does not show this branch is merged, so I have to close it :/

@theScrabi I suggest that you make the changes and push to the branch of the fork instead. That way, we could review them before it lands on the branch (even more so now that you blocked the `dev` branch), and the status of the pull request would be the right one.

Somethingweirdhere commented:
The PR seems nice, I will check it out soon :)

@mauriciocolli Thanks for looking over it. @theScrabi has made some changes in 4a19c78fa5fe8d44a10b15adf7158439a5dd56ec, so maybe some thing were fixed here. If I may ask, what does blocking the dev branch mean?

mauriciocolli commented:
I took a look and here's some things I think we have to fix:
- It would be easier if the persistence happened in a centralized manner (abstracted away from the usage). Currently, the saving/loading is spread across different classes, a maintenance hell.
    - Also, hardcoding preferences key across its usages.
- Saving/loading mechanism isn't too reliable, no tests.
    - Our priority is to make the app more reliable, this is the perfect time to introduce tests on every new feature (even more so in those that are easy to write one).
- Allow the usage of only one fixed fragment (e.g. subscriptions, feed, bookmarks), why would anyone need multiple exactly the same tabs?
- What if the user had a channel that was deleted in there? It will crash and block the user out of the application.
    - Though I think this was happening even before this PR.
- Some sections have weird indentation/code style.
- Not related to the code itself and a nitpick: some commits should have been squashed before merging.


I like this pull request, it just needed a little more polishing before it could make it to release.

If nobody have the time to open another one to fix these problems, I suggest we revert this pull request and open another to fix them and give it a proper review. It seems like there was a UI prioritization and the implementation was "as long as it works" in this one.

Somethingweirdhere commented:
> What if the user had a channel that was deleted in there? It will crash and block the user out of the application.

You are right, and it also happened before. This is probably the biggest issue here :/

> Our priority is to make the app more reliable, this is the perfect time to introduce tests on every new feature (even more so in those that are easy to write one).

I assume you mean [unit tests?](https://developer.android.com/training/testing/unit-testing/) Good point, I wanted to take a look at them, how they work and when/why to use them. If you don't mind a question: How can the saving/loading fail? And how can unit testing help? Just by detecting mistakes in the code so we can fix them?

> Allow the usage of only one fixed fragment (e.g. subscriptions, feed, bookmarks), why would anyone need multiple exactly the same tabs?

To be honest I don't know a situation where an user would want to have multiple of them, but I don't see any harm in it. If someone wants it this way, he can do it, and if he doesn't, he doesn't have to set it up this way.

> It would be easier if the persistence happened in a centralized manner (abstracted away from the usage). Currently, the saving/loading is spread across different classes, a maintenance hell.

After googling, I'm still not sure what that means, sorry. Do you mean that there should be a dedicated class that handles saving and loading data and provides functions for the other classes? Similarly to how it works with SQL?

> Also, hardcoding preferences key across its usages.

You are right, sorry for this.

> Some sections have weird indentation/code style.
> It seems like there was a UI prioritization and the implementation was "as long as it works" in this one.

You are definitely right here. I learned coding by "I want to do this, let's see what google says", and this project is my first confrontation with best practices. I will happily take a look at everything that's weird/wrong here to learn from it. I'm glad I learned why inheritance is useful, and a lot more things, and there are still things I'm wondering about. Everything is interesting, but I only know so much of best practices because I never had any education that relates to coding.

Anyways, thanks for looking over everything. I'm not sure what the right thing to do now is, but as soon as it's decided I'm happy to help out again, as well as I can.

mauriciocolli commented:
> I assume you mean unit tests? Good point, I wanted to take a look at them, how they work and when/why to use them. 

Yes, the main benefits from tests: confidence in your code. For example, when someone does some changes in some tested part, you can be pretty certain that it doesn't bring regressions or other side effects.

> How can the saving/loading fail? And how can unit testing help? Just by detecting mistakes in the code so we can fix them?

Pretty much. And it's not about how it will fail, it's about reducing the chances of it happening by covering several cases/situations/states that the it can be put through in a repeatable way.

---

I'll open a pull request later fixing some issues (hopefully before the release).
